CONTRIBUTING.md

```diff
@@ -0,0 +1,181 @@
+# Contribute to the secrets of sonar-text
+
+## Development environment
+
+1. Set up a Java development environment, or use Docker, to be ready to perform commands such as those specified in the [README.md](README.md)
+2. Set up an extension that validates YAML files, such as [YAML by RedHat](https://github.com/redhat-developer/vscode-yaml)
+3. Set up this [json-schema file](https://github.com/SonarSource/sonar-text/blob/master/sonar-text-plugin/src/main/resources/org/sonar/plugins/secrets/configuration/specifications/specification-json-schema.json) for your validation tooling
+4. Take a look at [the existing rules](https://github.com/SonarSource/sonar-text/tree/master/sonar-text-plugin/src/main/resources/org/sonar/plugins/secrets/configuration)
+5. Familiarize yourself with [SonarSource/rspec](https://github.com/SonarSource/rspec)
+
+## Adding a secret
+
+1. Write the yaml file
+2. Validates the yaml against the schema file
+3. use `secretSpecificationInclusionGenerator.sh <ruleApiFileName>`
+  1. It will generate the java files that will glue the yaml file to the code base
+  2. If the Rspec is already available, it will also generate additional rspec data
+4. Test the code:
+  1. If rspec data is available, do `mvn clean test`
+  2. Else, do `mvn test -Dtest=FakecloudCheckTest` and replace this test name by the one generated previously 
+
+## Common errors
+
+- Modifying code from the GitHub UI and not validating it with the schema
+  - It will raise tons of exceptions, and `"DeserializationException: Could not load specification from file: slack.yaml"]`  
+- Mismatch from what's expected in `rules[X].examples[X].match` and what's raised  (if `containsSecret` is set to true):
+  - `java.lang.AssertionError: actual is not empty while group of values to look for is.`
+- If `containsSecret` is set to false but something is still raised
+  - `Expecting empty but was: [DefaultIssue[ruleKey=secrets:SXXXXX,gap=<null>,overriddenSeverity=<null>,quickFixAvailable=false,ruleDescriptionContextKey=<null>,codeVariants=<null>,...,saved=true]]`
+- Creating overly greedy regex:
+  - The Some projects in the validation phase might not be analyzed because of time of scan
+## code structure for secret detection
```

Comment:
Add Jamie's categorization post

```suggestion
  - The Some projects in the validation phase might not be analyzed because of time of scan
  
### Not all secrets are created equal
We discovered that secrets generally fit into one of 3 categories:

1. Those that follow a pattern and contain identifiable sequences. For example, SonarQube API tokens always start with a known prefix (sqa_, sqp_ or squ_) that’s followed by 40 hexadecimal characters.
2. Those that follow a pattern, but which can only be identified by looking at the surrounding text. For example, AWS secret keys are always 40 base64 characters long, but only the surrounding text can identify whether they’re used with AWS.
3. Those that follow no fixed pattern, and which are only identifiable based on the surrounding text. For example, a PostgreSQL database password is user-chosen and must be set into a named environment variable before calling the psql command-line client.

As you go down the list, the secrets get harder to detect without also raising false positives. Category #3 is particularly difficult because the suspected secret could be a dummy value (e.g. [password]) or a variable that’s replaced at runtime.

A lot of tuning is needed in order to identify and exclude values that are common false positives.  
  
## code structure for secret detection
```